modules/gapi/test/common/gapi_core_tests_inl.hpp

```diff
@@ -1377,6 +1386,171 @@ TEST_P(NormalizeTest, Test)
+    std::vector<int> labels_gapi{};
+    initPointsVectorRandU(quantity, in_vector);
+    // G-API code //////////////////////////////////////////////////////////////
+    cv::GArray<cv::Point2f> in, centers;
+    cv::GArray<int> labels, gLabelsIn(std::vector<int>{});
+    cv::GOpaque<double> compactness;
+    std::tie(compactness, labels, centers) =
+        cv::gapi::kmeans(in, K, gLabelsIn, criteria, attempts, flags);
+    cv::GComputation c(cv::GIn(in), cv::GOut(compactness, labels, centers));
+    c.apply(cv::gin(in_vector), cv::gout(compact_gapi, labels_gapi, centers_gapi), getCompileArgs());
+    // Validation //////////////////////////////////////////////////////////////
+    {
+        EXPECT_GE(compact_gapi, 0.);
+        EXPECT_EQ(labels_gapi.size(), static_cast<size_t>(quantity));
+        EXPECT_EQ(centers_gapi.size(), static_cast<size_t>(K));
+    }
+}
+
+TEST_P(KMeans2DInitTest, AccuracyTest)
+{
+    const int quantity = sz.height;
+    const cv::TermCriteria criteria(TermCriteria::MAX_ITER + TermCriteria::EPS, 30, 0);
+    const int attempts = 1;
+    double compact_ocv = -1., compact_gapi = -1.;
+    std::vector<int> in_labels(quantity);
+    std::vector<int> labels_ocv{}, labels_gapi{};
+    std::vector<cv::Point2f> in_vector{}, centers_ocv{}, centers_gapi{};
+    initPointsVectorRandU(quantity, in_vector);
+    cv::randu(in_labels, 0, K);
+    labels_ocv = in_labels;
+    // G-API code //////////////////////////////////////////////////////////////
+    cv::GArray<cv::Point2f> in, centers;
+    cv::GArray<int> gLabelsIn, gLabelsOut;
+    cv::GOpaque<double> compactness;
+    std::tie(compactness, gLabelsOut, centers) =
+        cv::gapi::kmeans(in, K, gLabelsIn, criteria, attempts, flags);
+    cv::GComputation c(cv::GIn(in, gLabelsIn), cv::GOut(compactness, gLabelsOut, centers));
+    c.apply(cv::gin(in_vector, in_labels), cv::gout(compact_gapi, labels_gapi, centers_gapi),
+            getCompileArgs());
+    // OpenCV code /////////////////////////////////////////////////////////////
+    compact_ocv = cv::kmeans(in_vector, K, labels_ocv, criteria, attempts, flags, centers_ocv);
+    // Comparison //////////////////////////////////////////////////////////////
+    {
+        EXPECT_TRUE(compact_gapi == compact_ocv);
+        EXPECT_TRUE(compareVectorsAbsExact(labels_gapi, labels_ocv));
+        EXPECT_TRUE(compareVectorsAbsExact(centers_gapi, centers_ocv));
+    }
+}
+
+TEST_P(KMeans3DNoInitTest, AccuracyTest)
+{
+    const int quantity = sz.height;
+    const cv::TermCriteria criteria(TermCriteria::MAX_ITER + TermCriteria::EPS, 30, 0);
+    const int attempts = 1;
+    double compact_gapi = -1.;
+    std::vector<cv::Point3f> in_vector{}, centers_gapi{};
+    std::vector<int> labels_gapi{};
+    initPointsVectorRandU(quantity, in_vector);
+    // G-API code //////////////////////////////////////////////////////////////
+    cv::GArray<cv::Point3f> in, centers;
+    cv::GArray<int> labels, gLabelsIn(std::vector<int>{});
```

Comment:
here you have `labels` name, but in the test above it is a `gLabelsOut`. Why?
Also, I suggest to split definitions of the variables with the different purposes to the separate lines (for example, for input and output variables) or, if you think that this is extra, just give speaking names to the variables: 

 `cv::GArray<cv::Point3f> in, outCenters;`
  `cv::GArray<int> inLabels(std::vector<int>{}), outLabels;`

Reply:
Yes, thank you for these comments! Sorry, should have unified the names earlier.
Reworked